### PR TITLE
NP-51239 Extend APA-formatting

### PR DESCRIPTION
**Pull request by iBuzza**
# Description

Link to Jira issue: https://sikt.atlassian.net/browse/NP-51262

_Note: There is qiote a lot of diff in this PR. Most of it is from refactoring, tests and JSdocs. The big thing to focus on here is the new dispatcher-pattern_

Extends the APA citation formatting introduced in NP-51261 to cover more publication types, and adds a few related fixes.

  - Refactored `format-apa.ts` into a dispatcher pattern (one formatter per type) and split it into separate files under `_components/citation-box/_utils/`
  - Added formatters for **Book**, **Report**, and **Book Chapter**
  - Added a **generic fallback** formatter, so any unhandled type still produces a citation rather than rendering nothing
  - Author names are now used as stored on the registration (no more last-name + initials splitting, since the boundary between given names, middle names, and surnames isn't reliably knowable)
  - Unit tests added for each new formatter (full + partial metadata) and for the handle fallback

  # How to test

  Affected part of the application: Landing Page (citation box at top of the Details tab, still gated behind the beta flag)

  Open registrations of different types and confirm:
  - Journal article, book, book chapter, and report show a type-specific APA citation
  - Other types (e.g. conference lecture, degree) show a generic citation rather than nothing
  - A registration with a handle but no DOI shows the handle in the citation
  - Publisher name appears for books and reports

# Checklist

Ensure that the changes are aligned with the expectations of PO/designer before marking the PR as ready for review.

Also ensure that the following criterias are met:

- [x] The changes are working as expected
- [x] The changes are tested OK for different screen sizes
- [x] The changes are tested OK for a11y
- [x] Interactive elements have data-testids
- [x] I have done a QA of my own changes

_Note: some of these criterias may not always be relevant, and can simply be marked as completed._

## Release Notes

* **New Features**
  * Extended citation functionality to support books, book chapters, and reports alongside existing journal article citations.
  * Publisher and institution information is now included in all formatted citations.

* **Refactor**
  * Citation formatting system restructured to enable support for multiple publication types.

* **Tests**
  * Significantly expanded test coverage for multiple publication types and edge cases with incomplete metadata.

[Review Change Stack](https://app.coderabbit.ai/change-stack/BIBSYSDEV/NVA-Frontend/pull/8552?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/pages/public_registration/details-panel/_components/citation-box/_utils/formatter.types.ts
Comment (denektenina) [inline src/pages/public_registration/details-panel/_components/citation-box/_utils/formatter.types.ts]: Fin filplassering fra og med `details-panel`, men kanskje `details-panel` burde ligge under `registration-landing-page` istedet?

File: src/pages/public_registration/details-panel/_components/citation-box/_utils/citation-helpers.ts
Comment (denektenina) [inline src/pages/public_registration/details-panel/_components/citation-box/_utils/citation-helpers.ts:1]: Fin filplassering fra og med `details-panel,` men kanskje `details-panel` burde ligge under `registration-landing-page` istedet?

Comment (denektenina) [inline src/pages/public_registration/details-panel/_components/citation-box/_utils/citation-helpers.ts:1]: Utrolig behagelig å lese nå med kommentar som forklarer domenelogikk og if'ene tydelig viser hva som returneres i hvert tilfelle. Og veldig fin og ryddig gjenbruk av kode.

File: src/pages/public_registration/details-panel/_components/citation-box/_utils/format-apa.ts [hunk 1]
```diff
@@ -1,134 +1,47 @@
...
-  const pid = reference?.doi || registration.doi || registration.handle || '';
-
-  const authorYearSegment = formatAuthorYearSegment(authors, year);
-  const titleSegment = title ? `${title}.` : '';
-  const journalSegment = formatJournalSegment(journalTitle, formatVolumeIssue(volume, issue), pages);
-
-  return joinNonEmpty([authorYearSegment, titleSegment, journalSegment, pid]);
+  const instanceType = registration.entityDescription?.reference?.publicationInstance?.type;
+  const formatter = (instanceType && formattersByInstanceType[instanceType]) || formatGeneric;
+  return formatter(registration, options);
```
Comment (denektenina) [inline src/pages/public_registration/details-panel/_components/citation-box/_utils/format-apa.ts:46]: Elegant!

File: src/pages/public_registration/details-panel/_components/citation-box/_utils/formatters/format-chapter.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import { Registration } from '../../../../../../../types/registration.types';
+import { ChapterPublicationInstance } from '../../../../../../../types/publication_types/chapterRegistration.types';
+import { formatAuthorYearSegment, formatPages, joinNonEmpty, normalizeBaseFields } from '../citation-helpers';
+import { FormatAPAOptions, Formatter } from '../formatter.types';
+
+const formatChapterContainerSegment = (editors: string, bookTitle: string, pages: string): string => {
+  // The "In ..." segment is anchored on the parent book title; without it the editor/pages are meaningless.
+  if (!bookTitle) return '';
+  const editorPart = editors ? `${editors} (Ed.), ` : '';
```
Comment (denektenina) [inline src/pages/public_registration/details-panel/_components/citation-box/_utils/formatters/format-chapter.ts:9]: Ser ut til at i APA-stilformattering skal ha (Eds.) i flertall når det er flere editors.

**Review by denektenina — DISMISSED** [review]
Veldig fin og ryddig PR. Elegant løsning av et stort domene med veldig mange ulike cases og bra bruk av helpers for felles kode og dedikerte formatterere for de mange ulike casene. 

Kan approve denne uendret, men det er ett område som jeg vil tro bør fikses senere om ikke nå, og det er at det er hardkodet (Ed.) bak redaktører uavhengig av om det er én eller flere. Etter hva jeg så da jeg googlet APA-formatet virker det som det skal stå (Eds.) når det er flere redaktører.

File: src/pages/public_registration/details-panel/_components/citation-box/CitationBox.tsx [hunk 3]
```diff
@@ -13,7 +14,13 @@ interface CitationBoxProps {
...
   const { t } = useTranslation();
   const { journalName } = useJournalSeoData(registration);
-  const citation = formatAPA(registration, { journalName });
+
+  const publicationContext = registration.entityDescription?.reference?.publicationContext;
+  const publisherId =
+    (publicationContext && 'publisher' in publicationContext && publicationContext.publisher?.id) || '';
+  const publisherName = useFetchPublisherFromId(publisherId).data?.name ?? '';
+
+  const citation = formatAPA(registration, { journalName, publisherName });
```
Comment (denektenina) [inline src/pages/public_registration/details-panel/_components/citation-box/CitationBox.tsx]: Antar at editors og bookTitle skal sendes inn en en senere iterasjon?

File: src/pages/public_registration/details-panel/_components/citation-box/_utils/citation-helpers.ts
Comment (iBuzza) [reply src/pages/public_registration/details-panel/_components/citation-box/_utils/citation-helpers.ts:1]: Tenkte jeg skulle ta en runde på hele `action-panel`, `details-panel` og `log-panel` når citation er på plass. Lettere å se en felles context når alt flyttes samtidig

File: src/pages/public_registration/details-panel/_components/citation-box/_utils/formatters/format-chapter.ts (hunk 2 above)
Comment (iBuzza) [reply src/pages/public_registration/details-panel/_components/citation-box/_utils/formatters/format-chapter.ts:9]: Etter avklaring me PO så skal vi ha `(Ed.)` bak hver enkelt editor, siden de ikke nødvendigvis dukker opp i rekkefølge. Har en egen task på authorList formattering, så tar den der.

File: src/pages/public_registration/details-panel/_components/citation-box/CitationBox.tsx (hunk 3 above)
Comment (iBuzza) [reply src/pages/public_registration/details-panel/_components/citation-box/CitationBox.tsx]: Se kommentar over for editors. trodde av en eller annen grunn at vi ikke hadde en hook for å hente bookTitle, så avventet med den for å unngå større PR. Men fant den med Claude 🎉 Lagt til

File: src/pages/public_registration/details-panel/_components/citation-box/_utils/citation-helpers.ts [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+  if (names.length === 0) {
+    return '';
+  }
+  if (names.length === 1) {
+    return names[0];
+  }
+  if (names.length === 2) {
+    return `${names[0]}, & ${names[1]}`;
+  }
```
Comment (joachimjorgensen) [inline src/pages/public_registration/details-panel/_components/citation-box/_utils/citation-helpers.ts:25]: Can be removed?
```suggestion

```

**Review by joachimjorgensen — APPROVED** [review]
